### DISCUSSION TITLE
Using 8 relay HAT

### DISCUSSION THREAD

**Discussion by t4islander** [Dan-in-CA/SIP#257] [Q&A]
Hello, starting with total newbie to this :)

I have a Raspberry PI that I have connected to a 8 relay output board, its a HAT I think because I have landed the RPi directly to the board. I'm able to toggle relays on and off using some pretty generic Python scripts but when I installed SIP and run the program none of the relays are activating. I have been using the following code to test with in Python as an example:

import RPi.GPIO as GPIO
import time

GPIO.setmode(GPIO.BCM)
GPIO.setwarnings(False)

GPIO.setup(26, GPIO.OUT)

GPIO.output(26, GPIO.HIGH) #relay off
time.sleep(.5)
GPIO.output(26, GPIO.LOW) #relay on
time.sleep(10)
GPIO.output(26, GPIO.HIGH) #relay off 

Is there a configuration somewhere I can change to use this board? I have GPIO PINs 5,6,13,16,19,20,21,26  that are working in my own Python but would rather not try to invent something that already exists and is really well built!

**Comment by t4islander** [comment]
ok, so quick update - I installed the relay_16 plugin and now when I run manually SO6 fires relay 4, SO7 fires relay 6, and SO8 fires relay 7 on my board. Is there somewhere I can update the GPIO mappings? I looked at the relay_16.py file and see the mapping there. I was thinking I could list the GPIO PINs I'm using there but I can't update the file, and don't know if I should. Any help much appreciated :)

**Reply by Dan-in-CA** [reply]
Welcome,
What type of relay hat are you using? Is there some online documentation for it?

There are a couple of possible ways to get it working for you.
Updating the pin mapping could work but that would need to be done in SIP/gpio_pins.py starting at line 106. The main problem with that is if you update SIP in the future you would need to re-map the pins again.

If I could learn more about the hat you are using it may be possible to come up with an easier solution.

I am sure we can get it working.

Dan

**Reply by t4islander** [reply]
Thanks for the response and offer for help, really appreciate it. There was literally no documentation that came with the board, but its an 8 relay board called 'Relay Board B for Raspberry' on one sticker then another says 'Raspberry Pi Expansion board 8'. Waveshare is the manufacturer or at least the box is came in was labeled that. I purchased it off of Amazon.

**Reply by Noschvie** [reply]
https://www.waveshare.com/w/upload/6/6f/RPi_Relay_Board_%28B%29_User_Manual_EN.pdf

**Reply by Dan-in-CA** [reply]
OK,
After giving this a bit more thought you will need to edit the pin mapping in relay_16.py as you first mentioned. This will avoid any problem updating SIP in the future.

The relay_16 plugin does not use all available GPIO pins. It avoids the low number pins so that things like a real-time clock can use them. By editing the mapping in relay_16.py you will be able to use the pins your relay hat is designed to use.
It looks like the pin numbers you listed in your first post are BMC numbering (5,6,13,16,19,20,21,26). you will need to translate those numbers to the board numbering that SIP uses  (29.31,33,36,35,38,40,37). I think this will work but I have not tested it.
Try using that numbering in relay_16.py for the first 8 values in the list. A good reference for the numbering is:
https://pinout.xyz/#

**Reply by t4islander** [reply]
IT WORKS! Awesome, I'm excited to get these relays connected to the sprinkler system and start being efficient :). Really truly appreciate the help here, great program and forum you have here.

**Reply by t4islander** [reply]
Ok, trying to incorporate the Weather based plug in to moderate water use I am getting the following error from the service. Any ideas on how to overcome?

Weather-based water level encountered error:
Traceback (most recent call last):
  File "/home/pi/SIP/plugins/weather_level_adj.py", line 123, in run
    today = today_info(self, options)
  File "/home/pi/SIP/plugins/weather_level_adj.py", line 627, in today_info
    data = get_data(name, request, u"weather", options)
  File "/home/pi/SIP/plugins/weather_level_adj.py", line 495, in get_data
    req = urlopen(url + u"&appid=" + options[u"apikey"])
  File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
HTTPError: HTTP Error 400: Bad Request

**Reply by Dan-in-CA** [reply]
That plugin was modified by several contributors. I am testing but this may take a while to debug.
I will let you know how it goes. 

For now it would be best to disable it and get the rest of your system running.

**Reply by Dan-in-CA** [reply]
I was able to run the plugin successfully.

When I deleted the Openweather API Key I got a 401 Unauthorized error.
When I deleted the location info I got a 400 Bad request error like you got.

Make sure the location info is formatted as described in the Help page.

Dan

**Reply by t4islander** [reply]
Thanks Dan - again really appreciate the help with this. I went ahead and connected up the relays to the sprinkler valves and it's working like a champ. I'll double check my settings and continue to troubleshoot.

**Reply by t4islander** [reply]
Update: Well I went it to check the settings and the call was successful! Looks like I'm on the path to green lawns! Thanks again for the help and responses, have a great day.
